### PR TITLE
FIX-#4449 Drain the call queue before waiting on result in benchmark mode.

### LINKED ISSUE #4449: Benchmark Mode blocks on computation so all computation occurs on one node
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **Modin version** (`modin.__version__`):
- **Python version**:
- **Code we can use to reproduce**:

### Describe the problem

Since Benchmark Mode calls `wait` on each partition sequentially, computation that can be done in parallel ends up happening sequentially on a single node. We need to instead collect all the OID's of all partitions, and wait on all of them together - that way each of the tasks is kicked off in parallel, and computation happens in parallel.

### PR DESCRIPTION
**Pull request by mvashishtha**
## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4449 
- [ ] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 
- [x] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 1]
```diff
@@ -60,6 +60,13 @@ def wait(*args, **kwargs):
                 partitions = result[0]
             else:
                 partitions = result
+            # When partitions have a deferred call queue, calling
+            # parittion.wait() on each partition serially will serially kick
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:64]:
```suggestion
            # partition.wait() on each partition serially will serially kick
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 2]
```diff
@@ -60,6 +60,13 @@ def wait(*args, **kwargs):
                 partitions = result[0]
             else:
                 partitions = result
+            # When partitions have a deferred call queue, calling
+            # partition.wait() on each partition serially will serially kick
+            # off each deferred computation and wait for each partition to
+            # finish before kicking off the next one. Instead, we want to
+            # serially kick off all the deferred computations so that they can
+            # all run aynschronously, then serially wait on all the results.
+            [part.drain_call_queue() for row in partitions for part in row]
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:69]:
```suggestion
            [part.drain_call_queue() for part in partitions.flatten()]
```

**Review by anmyachev — DISMISSED** [review]

**Review by anmyachev — APPROVED** [review]

**Comment by YarShev** [comment]
It seems to me that @RehanSD meant something like this when creating the issue.
```python
ray.wait([part.oid for part in partitions], num_returns=len(partitions))
```
@RehanSD, please correct me if I am wrong.

**Comment by mvashishtha** [comment]
> It seems to me that @RehanSD meant something like this when creating the issue.
> 
> ```python
> ray.wait([part.oid for part in partitions], num_returns=len(partitions))
> ```
> 
> @RehanSD, please correct me if I am wrong.

@YarShev that won't fix the bug. e.g. if we have a transpose on the call queue, `ray.wait` will return immediately for each partition's `oid`. We need to drain the call queue before waiting.

**Comment by YarShev** [comment]
@mvashishtha, yes, I understand. I am saying about waiting on partition ids after we drain the call queue.

**Comment by mvashishtha** [comment]
> @mvashishtha, yes, I understand. I am saying about waiting on partition ids after we drain the call queue.

@YarShev what's wrong with waiting the way we are doing now? We are not waiting in parallel, but any remote functions should be executing in parallel anyway. Doing a `ray wait` would let us wait for everything at once, but I don't expect it to save any significant time because the cost of iterating through the partition is so little.

**Comment by YarShev** [comment]
@mvashishtha, I would like to make sure if there is a difference between waiting on each partition separately and on all partitions at once both in a single node and in a cluster. @modin-project/modin-ray, could you shed a light on this?

**Comment by mvashishtha** [comment]
@YarShev Isn't that question separate from the problem this PR is addressing? Draining the call queue is something we have to do serially anyway.

**Comment by YarShev** [comment]
@mvashishtha, I think my question may be related to the problem this PR is addressing. I am wondering if there is a parallel wait on object references in case of a cluster?

**Comment by wuisawesome** [comment]
If I'm understanding this conversation correctly, `ray.wait(refs)` waits on all the refs in parallel. (Note `num_returns=1` by default so we have to wait for all of them to know which one finishes first).

**Comment by YarShev** [comment]
@wuisawesome, that is what I expected. Could you tell us about how the parallel wait works? Does the parallelization happen per workers or per nodes (plasma)?

**Review by devin-petersohn — APPROVED** [review]
@mvashishtha Makes sense, I thought this bug was an issue with synchronous execution.

LGTM

**Comment by YarShev** [comment]
Still, I think we should use a parallel wait from Ray (and probably from Dask) to check the performance correctly. We likely want to wait for partitions when all computations are complete in parallel. I guess we could add a method named `wait` to the partition manager where we wait for all partitions to be ready.

**Comment by mvashishtha** [comment]
> Still, I think we should use a parallel wait from Ray (and probably from Dask) to check the performance correctly. We likely want to wait for partitions when all computations are complete in parallel. I guess we could add a method named `wait` to the partition manager where we wait for all partitions to be ready.

@YarShev I still think this is a separate issue that is independent of this one. I have filed #4491 for it. This issue is about deferred tasks that are executed in sequence in benchmark mode, whereas #4491 is about more precisely waiting for all async computations in benchmark mode in general. Let's please keep this PR focused on #4449.

**Comment by YarShev** [comment]
> > Still, I think we should use a parallel wait from Ray (and probably from Dask) to check the performance correctly. We likely want to wait for partitions when all computations are complete in parallel. I guess we could add a method named `wait` to the partition manager where we wait for all partitions to be ready.
> 
> @YarShev I still think this is a separate issue that is independent of this one. I have filed #4491 for it. This issue is about deferred tasks that are executed in sequence in benchmark mode, whereas #4491 is about more precisely waiting for all async computations in benchmark mode in general. Let's please keep this PR focused on #4449.

@mvashishtha, ok

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 3]
```diff
@@ -60,9 +60,18 @@ def wait(*args, **kwargs):
                 partitions = result[0]
             else:
                 partitions = result
+            # When partitions have a deferred call queue, calling
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:63]: I think this comment should be rephrased taking into account these discussions - [link1](https://github.com/modin-project/modin/pull/4472#issuecomment-1136042874) and [link2](https://github.com/modin-project/modin/pull/4472#issuecomment-1136219274).

Comment (mvashishtha) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:63]: What would you change it to? It seems like as is, it accurately explains why the `drain_call_queue` is there.

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:63]: I read again this and probably would change the last part a bit. Left a suggestion.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 4]
```diff
@@ -60,9 +60,18 @@ def wait(*args, **kwargs):
...
+            # partition.wait() on each partition serially will serially kick
+            # off each deferred computation and wait for each partition to
+            # finish before kicking off the next one. Instead, we want to
+            # serially kick off all the deferred computations so that they can
+            # all run aynschronously, then serially wait on all the results.
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:68]:
```suggestion
            # all run asynchronously, then wait on all the results in parallel.
```

Comment (mvashishtha) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:68]: I wanted to clarify that while the computations are happening in parallel, we wait on them in serial (for now). What about this?

```suggestion
            # all run asynchronously in parallel, then wait serially on all the results.
```

Comment (YarShev) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:68]: The sentence begins with `we want to ...` while actuaclly we do not want to wait serially and will change this in the future. Let's just not use `in parallel` here. Corrected my suggestion.

**Review by YarShev — APPROVED** [review]

**Pull request merged**